src/dataflow-types/src/client/controller/storage.rs

```diff
@@ -99,12 +97,19 @@ impl<'a, C: Client<T>, T: Timestamp + Lattice> StorageController<'a, C, T> {
         Ok(())
     }
     pub async fn drop_sources(&mut self, identifiers: Vec<GlobalId>) {
+        // Dropping a source is equivalent to releasing any implicit capability.
+        // TODO(mcsherry): hold back this announcement until all capabilities are released.
         for id in identifiers.iter() {
-            if !self.storage.source_descriptions.contains_key(id) {
-                // This isn't an unrecoverable error, just .. probably wrong.
-                error!("Source id {} dropped without first being created", id);
-            } else {
-                self.storage.source_descriptions.insert(*id, None);
+            match self.collection_mut(*id) {
+                Ok(collection) => {
+                    // Apply the updates but ignore the results for now.
+                    // TODO(mcsherry): observe the results and allow compaction.
+                    let _ = collection.capability_downgrade(Antichain::new());
+                }
+                Err(_identifier_missing) => {
```

Comment:
Should we match the specific variant here and put a `unreachable!()` branch if that is the only thing that can happen?

Reply:
That's a good point. I think this is still in flux a bit, trying to sort out which things are errors and which we should absorb. I'll try and clean this up in the next revision!